File: 07-reference-implementation/sdk/python/examples/agent_payment.py
"""
Example: AI Agent executing a payment using ACP (ACP-HP-1.0).

Demonstrates the full verification flow:
  1. Generate agent identity (Ed25519)
  2. Issue + sign a capability token (in production: done by institutional issuer)
  3. Connect to an ACP server and call client.verify()

Requirements:
    pip install acp-sdk           # or: pip install cryptography
    Optional: set ACP_AGENT_SEED  # 32 bytes hex — for deterministic identity
    Optional: set ACP_SERVER_URL  # default: http://localhost:8080

Run:
    python examples/agent_payment.py

To test with the Go reference server:
    cd acp-go && go build -o acp-server.exe ./cmd/acp-server
    ./acp-server.exe &
    python examples/agent_payment.py
"""
from __future__ import annotations

import json
import os
import secrets
import time

from acp.identity import AgentIdentity
from acp.signer import ACPSigner
from acp.client import ACPClient, ACPError


def _load_identity() -> AgentIdentity:
    """Load from ACP_AGENT_SEED env var, or generate a new ephemeral identity."""
    seed_hex = os.getenv("ACP_AGENT_SEED", "")
    if seed_hex:
        return AgentIdentity.from_private_bytes(bytes.fromhex(seed_hex))
    print("[INFO] ACP_AGENT_SEED not set — generating ephemeral identity")
    return AgentIdentity.generate()


def _build_payment_token(
    issuer: AgentIdentity,
    subject: AgentIdentity,
) -> dict:
    """
    Build and sign a financial.payment capability token.

    NOTE: In production, the institutional issuer creates and signs tokens.
          This local demo uses a freshly generated issuer for illustration.
    """
    now = int(time.time())
    nonce = secrets.token_urlsafe(16)

    capability = {
        "ver": "1.0",
        "jti": secrets.token_urlsafe(16),
        "iss": issuer.did,
        "sub": subject.agent_id,
        "iat": now,
        "exp": now + 3600,
        "nonce": nonce,
        "capabilities": ["acp:cap:financial.payment"],
        "constraints": {
            "max_amount_usd": 10000,
            "allowed_accounts": ["ACC-001", "ACC-002"],
        },
    }

    signer = ACPSigner(issuer)
    return signer.sign_capability(capability)


def main() -> None:
    print("=== ACP Python SDK — Agent Payment Example (ACP-HP-1.0) ===\n")

    # Step 1 — Agent identity
    agent = _load_identity()
    signer = ACPSigner(agent)
    print(f"Agent ID  : {agent.agent_id}")
    print(f"Agent DID : {agent.did}")
    print(f"Pubkey    : {agent.public_key_bytes.hex()[:24]}...\n")

    # Step 2 — Issuer (separate identity; in production: institutional system)
    issuer = AgentIdentity.generate()
    print(f"Issuer DID: {issuer.did}\n")

    # Step 3 — Signed capability token
    token = _build_payment_token(issuer, agent)
    print("Capability token (summary):")
    print(f"  jti         : {token['jti']}")
    print(f"  capabilities: {token['capabilities']}")
    print(f"  exp         : {token['exp']} (now+1h)")
    print(f"  signature   : {token['proof']['signature'][:32]}...\n")

    # Step 4 — Verify signature locally (no server needed)
    is_valid = ACPSigner.verify_capability(token, issuer.public_key_bytes)
    print(f"Local signature valid : {is_valid}")

    # Step 5 — Connect to ACP server and run full PoP handshake
    server_url = os.getenv("ACP_SERVER_URL", "http://localhost:8080")
    client = ACPClient(
        server_url=server_url,
        identity=agent,
        signer=signer,
    )

    print(f"\nConnecting to ACP server: {server_url}")
    try:
        health = client.health()
        print(f"Server health: {health}\n")
    except ACPError as e:
        print(f"[Server not reachable — {e}]")
        print("Continuing with local verification only.\n")
        _show_offline_demo(agent, signer, token)
        return

    # Full ACP-HP-1.0 verification (challenge → PoP → verify)
    print("Running full ACP-HP-1.0 verification flow...")
    try:
        result = client.verify(
            capability_token=token,
            requested_capability="acp:cap:financial.payment",
            requested_resource="org.banco-soberano/accounts/ACC-001",
        )
        print(f"Decision  : {result.get('decision')}")
        print(f"Full response: {json.dumps(result, indent=2)}")
    except ACPError as e:
        print(f"ACP error (status={e.status_code}): {e}")


def _show_offline_demo(
    agent: AgentIdentity,
    signer: ACPSigner,
    token: dict,
) -> None:
    """Show what the PoP payload would look like without a live server."""
    import base64
    import hashlib

    fake_challenge = base64.urlsafe_b64encode(b"demo-challenge-nonce").rstrip(b"=").decode()
    capability_id = "acp:cap:financial.payment"
    binding = f"{fake_challenge}.{agent.agent_id}.{capability_id}".encode()
    digest = hashlib.sha256(binding).digest()
    sig_bytes = signer.sign_bytes(digest)
    sig_b64 = base64.urlsafe_b64encode(sig_bytes).rstrip(b"=").decode()

    print("--- Offline PoP demo ---")
    print(f"Challenge : {fake_challenge}")
    print(f"Agent ID  : {agent.agent_id}")
    print(f"Capability: {capability_id}")
    print(f"PoP sig   : {sig_b64[:40]}...")
    print("\nTo run the full flow, start the Go reference server:")
    print("  cd acp-go && go build -o acp-server.exe ./cmd/acp-server && ./acp-server.exe")
    print("  ACP_SERVER_URL=http://localhost:8080 python examples/agent_payment.py")


if __name__ == "__main__":
    main()

File: impl/typescript/examples/agent_payment.ts
/**
 * Example: AI Agent executing a payment using ACP TypeScript SDK.
 *
 * Demonstrates the complete ACP-HP-1.0 flow for an autonomous agent:
 *   1. Load/generate Ed25519 identity
 *   2. Create a capability token (issued by institutional authority)
 *   3. Execute the payment via ACPClient (auto handles challenge/PoP)
 *
 * Environment variables:
 *   ACP_AGENT_SEED   64-char hex-encoded 32-byte Ed25519 seed (optional)
 *   ACP_SERVER_URL   ACP server base URL (default: http://localhost:8080)
 *
 * Run (requires Go server):
 *   cd acp-go && ACP_INSTITUTION_PUBLIC_KEY=<pubkey> go run ./cmd/acp-server &
 *   ACP_SERVER_URL=http://localhost:8080 npx tsx examples/agent_payment.ts
 */

import { ACPIdentity, ACPClient, signToken } from "../src/index.js";
import type { CapabilityToken } from "../src/index.js";

// ─── Identity ─────────────────────────────────────────────────────────────────

function loadOrGenerateIdentity(): ACPIdentity {
  const seedHex = process.env.ACP_AGENT_SEED;
  if (seedHex && seedHex.length === 64) {
    console.log("[ACP] Loading identity from ACP_AGENT_SEED");
    return ACPIdentity.fromHex(seedHex);
  }
  console.warn("[ACP] ACP_AGENT_SEED not set — generating ephemeral identity");
  return ACPIdentity.generate();
}

// ─── Token Factory ────────────────────────────────────────────────────────────

/**
 * Create a test capability token signed by the issuer.
 *
 * In production:
 *   - The institutional issuer signs tokens via their secure HSM/KMS
 *   - The agent NEVER signs its own tokens
 *   - This is ONLY for local testing/demo purposes
 */
function createTestToken(
  issuerIdentity: ACPIdentity,
  agentIdentity: ACPIdentity
): string {
  const now = Math.floor(Date.now() / 1000);

  // Generate random 128-bit nonce.
  const nonceBytes = new Uint8Array(16);
  crypto.getRandomValues(nonceBytes);
  const nonce = Buffer.from(nonceBytes).toString("base64url");

  const payload: Omit<CapabilityToken, "sig"> = {
    ver: "1.0",
    iss: issuerIdentity.agentId,
    sub: agentIdentity.agentId,
    cap: ["acp:cap:financial.payment"],
    res: "org.banco-soberano/accounts/ACC-001",
    iat: now,
    exp: now + 3600,
    nonce,
    deleg: { allowed: false, max_depth: 0 },
    parent_hash: null,
    constraints: { max_amount_usd: 10000 },
    rev: {
      type: "endpoint",
      uri: "https://acp.banco-soberano.com/acp/v1/rev/check",
    },
  };

  const signed = signToken(payload, issuerIdentity);
  return JSON.stringify(signed);
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main(): Promise<void> {
  console.log("=== ACP TypeScript SDK — Agent Payment Example ===\n");

  // 1. Agent identity.
  const agent = loadOrGenerateIdentity();
  console.log(`Agent AgentID : ${agent.agentId}`);
  console.log(`Agent PubKey  : ${Buffer.from(agent.publicKey).toString("hex").slice(0, 16)}...\n`);

  // 2. Institutional issuer (separate keypair in production).
  const issuer = ACPIdentity.generate();
  console.log(`Issuer AgentID: ${issuer.agentId}`);
  console.log(`Issuer PubKey : ${Buffer.from(issuer.publicKey).toString("hex").slice(0, 16)}...\n`);

  // 3. Create capability token.
  const tokenJson = createTestToken(issuer, agent);
  const tokenPreview = tokenJson.slice(0, 100);
  console.log(`Token (first 100 chars): ${tokenPreview}...\n`);

  // 4. Initialize ACP client.
  const baseUrl = process.env.ACP_SERVER_URL ?? "http://localhost:8080";
  const client = new ACPClient(agent, { baseUrl, timeoutMs: 5000 });

  // 5. Register agent with server (for PoP verification).
  console.log("Registering agent with server...");
  try {
    const pubKeyB64 = Buffer.from(agent.publicKey).toString("base64url");
    const regResponse = await fetch(`${baseUrl}/acp/v1/register`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ agent_id: agent.agentId, public_key_hex: pubKeyB64 }),
    });
    const regData = await regResponse.json() as { ok: boolean };
    if (regData.ok) {
      console.log("Agent registered successfully\n");
    }
  } catch (err) {
    console.log(`[Expected without server] Registration: ${(err as Error).message}\n`);
  }

  // 6. Execute payment with full ACP-HP-1.0 handshake.
  const transferPayload = {
    to_account: "ACC-999",
    amount: 500,
    currency: "USD",
    memo: "Invoice payment #2024-001",
  };

  console.log(`Executing payment: ${JSON.stringify(transferPayload)}`);
  console.log(`Server          : ${baseUrl}\n`);

  try {
    const response = await client.execute({
      method: "POST",
      path: "/api/v1/payments/transfer",
      capabilityToken: tokenJson,
      payload: transferPayload,
    });

    console.log(`Response status: ${response.status}`);
    const text = await response.text();
    console.log(`Response body  : ${text.slice(0, 300)}`);
  } catch (err) {
    const e = err as Error;
    console.log(`[Expected in demo without configured server]`);
    console.log(`${e.constructor.name}: ${e.message}`);
    console.log("\n--- To run the full demo ---");
    console.log("1. cd acp-go");
    console.log(`2. ACP_INSTITUTION_PUBLIC_KEY=${Buffer.from(issuer.publicKey).toString("base64url")} go run ./cmd/acp-server`);
    console.log("3. ACP_SERVER_URL=http://localhost:8080 npx tsx examples/agent_payment.ts");
  }
}

main().catch(console.error);
